source/error.c

```diff
@@ -0,0 +1,135 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/error.h>
+#include <aws/common/common.h>
+#include <assert.h>
+
+AWS_THREAD_LOCAL int last_error = 0;
+
+static aws_error_handler global_handler = NULL;
+static void *global_error_context = NULL;
+
+AWS_THREAD_LOCAL aws_error_handler thread_handler = NULL;
+AWS_THREAD_LOCAL void *thread_handler_context = NULL;
+
+#ifndef AWS_MAX_ERROR_SLOTS
+#define AWS_MAX_ERROR_SLOTS 10
+#endif
+
+#define AWS_ERROR_SLOT_SIZE 1000
+
+static volatile const struct aws_error_info_list *error_slots[AWS_MAX_ERROR_SLOTS] = { 0 };
+
+int aws_last_error(void) {
+    return last_error;
+}
+
+static const struct aws_error_info *get_error_by_code(int err) {
+    if(err >= AWS_MAX_ERROR_SLOTS * AWS_ERROR_SLOT_SIZE) {
+        return NULL;
+    }
+
+    int slot_index = err / AWS_ERROR_SLOT_SIZE;
```

Comment:
We should make AWS_ERROR_SLOT_SIZE a power of two to avoid division.

Reply:
will do.

Reply:
crap forgot this one. one sec

Reply:
wait lets hold off on this one. for one, this isn't hit in high-perf code, and I like the predictable ranges we get with 1000.

Reply:
What about predictable ranges of `0x1000`?

Reply:
that's a good idea.

Reply:
let's do 0x400 = 1024